server/bin/pbench-audit-server.py

```diff
@@ -0,0 +1,852 @@
+        self.config = config
+
+        self.controllers = list()
+        self.things_to_check = {
+            "bad_controllers": list(),
+            "tarballs": DictOfList(),
+            "unexpected_symlinks": DictOfList(),
+            "unexpected_objects": DictOfList(),
+            "expected_objects": DictOfList(),
+            "unexpected_dirs": DictOfList(),
+            "non_prefixes": DictOfList(),
+            "wrong_prefixes": DictOfList(),
+            "non_prefix_dirs": DictOfList(),
+            "status_indicators": DictOfList(),
+        }
+
+    def add_controller(self, controller):
+        self.controllers.append(controller)
+
+    def add_bad_controller(self, controller):
+        self.things_to_check["bad_controllers"].append(controller)
+
+    def add_tarballs(self, controller, tarball):
+        self.things_to_check["tarballs"].add_entry(controller, tarball)
+
+    def add_unexpected_symlinks(self, controller, link):
+        self.things_to_check["unexpected_symlinks"].add_entry(controller, link)
+
+    def add_unexpected_objects(self, controller, obj):
+        self.things_to_check["unexpected_objects"].add_entry(controller, obj)
+
+    def add_unexpected_dirs(self, controller, dirs):
+        self.things_to_check["unexpected_dirs"].add_entry(controller, dirs)
+
+    def add_non_prefixes(self, controller, prefix):
+        self.things_to_check["non_prefixes"].add_entry(controller, prefix)
+
+    def add_wrong_prefixes(self, controller, prefix):
+        self.things_to_check["wrong_prefixes"].add_entry(controller, prefix)
+
+    def set_status(self, controller, subdir):
+        self.things_to_check["status_indicators"].add_entry(controller, subdir)
+
+    def dump(self, f):
+        cnt = 0
+        f.write(
+            "\nstart-{}: archive hierarchy: {}\n".format(self.config.TS[4:], self.path)
+        )
+        if self.things_to_check["bad_controllers"]:
+            f.write("\nBad Controllers:\n")
+            cnt += 1
+            for controller in sorted(self.things_to_check["bad_controllers"]):
+                f.write("\t* {}\n".format(controller))
+        for controller in sorted(self.controllers):
+            check = any(
+                [
+                    controller in self.things_to_check[k]._dict
+                    for k in self.things_to_check
+                    if k != "tarballs" and k != "bad_controllers"
+                ]
+            )
+            if check or controller not in self.things_to_check["tarballs"]._dict:
+                f.write("\nController: {}\n".format(controller))
+                if controller in self.things_to_check["unexpected_dirs"]._dict:
+                    f.write(
+                        "\t* Unexpected state directories found in this controller directory:\n"
+                    )
+                    cnt = self.output_format(f, controller, "unexpected_dirs", cnt)
+                if controller in self.things_to_check["status_indicators"]._dict:
```

Comment:
Ditto for the "status_indicators" entry: instead of treating it as a special case inside the loop, take it out of `things_to_check` and treat it separately. Then the only things left are all things that you treat the same way: if controller is in the dict, then write a message and output all the things that have been accumulated for that entry: you can do that with a simple loop!

Reply:
We can do that when the results that are being produced are finalized. then we can refactor only dump method the way we want